src/cli/rustup_mode.rs

```diff
@@ -1312,10 +1320,20 @@ fn toolchain_link(cfg: &Cfg, m: &ArgMatches<'_>) -> Result<utils::ExitCode> {
 }
 
 fn toolchain_remove(cfg: &mut Cfg, m: &ArgMatches<'_>) -> Result<utils::ExitCode> {
-    for toolchain in m.values_of("toolchain").unwrap() {
-        let toolchain = cfg.get_toolchain(toolchain, false)?;
-        toolchain.remove()?;
+    if m.is_present("pattern") {
+        for pattern_str in m.values_of("toolchain").unwrap() {
+            let pattern = Pattern::new(pattern_str)?;
+            for toolchain in cfg.get_toolchains_from_glob(pattern)? {
+                toolchain.remove()?;
+            }
```

Comment:
Should we print a message if the pattern didn't match any toolchains so the user notices the silent "failure"?

Reply:
Yes, if nothing was removed it should be an error, same as if you try to remove a toolchain which is not installed now.  However I remain interested in if there's a way to not need the `--pattern` flag at all.  Could we simply always treat the incoming `toolchain` arguments as patterns?

Reply:
See my comment: https://github.com/rust-lang/rustup/pull/2540#discussion_r513770900

Reply:
Let's collate conversation on this point over there.